### PR TITLE
Drop lerna

### PR DESCRIPTION
**Pull request by dapplion**
**Motivation**

lerna has been used in this monorepo for +3y (added in https://github.com/ChainSafe/lodestar/pull/324). While useful we have had to battle this tool more than treat it like a helpful item:
- https://github.com/ChainSafe/lodestar/pull/4052
- https://github.com/ChainSafe/lodestar/pull/4049
- https://github.com/ChainSafe/lodestar/pull/3990
- https://github.com/ChainSafe/lodestar/pull/3679
- https://github.com/ChainSafe/lodestar/pull/3637
- https://github.com/ChainSafe/lodestar/pull/3601
- https://github.com/ChainSafe/lodestar/pull/2656
- https://github.com/ChainSafe/lodestar/pull/2650
- https://github.com/ChainSafe/lodestar/pull/2649
- https://github.com/ChainSafe/lodestar/pull/1909
- https://github.com/ChainSafe/lodestar/pull/666

It also deprecates scripts introduced here, not used currently AFAIK
- https://github.com/ChainSafe/lodestar/pull/1142

**Description**

We use lerna for three actions:
- Run scripts in all packages: Now done `yarn workspaces run <cmd>`
- Bump version and dependencies of local packages: Now done with `scripts/release/set_version.mjs`
- Publish all packages: Now done `scripts/release/publish.mjs`

Please check the new scripts and you'll see they are really simple. I don't think they justify a tool so heavy and problematic like lerna. I don't guarantee this PR won't introduce bugs for each type of release, but fixing those should be trivial now that we control the tooling.

**Please note that we have dropped 2_300 lines of dependencies from yarn.lock :tada:**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,33 +5,40 @@
...
+    "packages/spec-test-util",
+    "packages/config",
+    "packages/api",
+    "packages/db",
+    "packages/state-transition",
+    "packages/light-client",
+    "packages/validator",
+    "packages/flare",
+    "packages/fork-choice",
+    "packages/beacon-node",
```
Comment (dapplion) [inline package.json:20]: This hardcoded list is sorted topologically. Yarn v1 does not support topological inference out of the box, so this does the trick. Long term we can switch to newer versions of yarn that do this automatically.

File: packages/api/package.json [hunk 2]
```diff
@@ -56,17 +56,11 @@
...
     "check-build": "node -e \"(async function() { await import('./lib/index.js') })()\"",
     "check-types": "tsc",
-    "coverage": "codecov -F lodestar-api",
     "lint": "eslint --color --ext .ts src/ test/",
-    "lint:fix": "yarn run lint --fix",
-    "pretest": "yarn run check-types",
-    "test": "yarn test:unit && yarn test:e2e",
     "test:unit": "nyc --cache-dir .nyc_output/.cache -e .ts mocha 'test/unit/**/*.test.ts'",
-    "check-readme": "typescript-docs-verifier"
+    "test:e2e": "echo 'n/a'"
```
Comment (dapplion) [inline packages/api/package.json:63]: yarn v1 `workspace run` util requires all packages to have the script declared

File: packages/api/package.json [hunk 3]
```diff
@@ -56,17 +56,11 @@
...
-    "build:release": "yarn clean && yarn run build && yarn run build:typedocs",
     "check-build": "node -e \"(async function() { await import('./lib/index.js') })()\"",
     "check-types": "tsc",
-    "coverage": "codecov -F lodestar-api",
     "lint": "eslint --color --ext .ts src/ test/",
-    "lint:fix": "yarn run lint --fix",
-    "pretest": "yarn run check-types",
-    "test": "yarn test:unit && yarn test:e2e",
     "test:unit": "nyc --cache-dir .nyc_output/.cache -e .ts mocha 'test/unit/**/*.test.ts'",
-    "check-readme": "typescript-docs-verifier"
```
Comment (dapplion) [inline packages/api/package.json:69]: CI only items like `check-readme`, `coverage` and `build:typedocs` and handled by other scripts only meant for CI

File: scripts/foreach.sh [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+WORKSPACE_DIR=./packages
+
+for d in $WORKSPACE_DIR/* ; do
+    (cd "$d" && eval "$@")
+done
```
Comment (dapplion) [inline scripts/foreach.sh:7]: `lerna exec <cmd>` implemented in 3 lines of bash

File: scripts/release/create_rc.mjs [hunk 5]
```diff
@@ -102,8 +102,8 @@ if (!(await confirm(`Do you want to create a release candidate for ${versionMMP}
 // Create a new release branch `rc/v1.1.0` at commit `9fceb02`
 shell(`git checkout -b ${rcBranchName} ${commit}`);
 
-// Set monorepo version to `1.1.0`
-shell(`lerna version ${packageVersion} --no-git-tag-version --force-publish --yes`);
+// Set monorepo version to `1.1.0`. Use tilde to apply only patch releases, not minor
+shell(`node scripts/release/set_version.mjs ${packageVersion} tilde`);
```
Comment (dapplion) [inline scripts/release/create_rc.mjs:106]: tilde here means inter-monorepo deps with be `~0.40.0` such that only patch releases can be installed in that range. `^0.40.0` applies to minor releases too.

File: scripts/release/publish.mjs [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+Usage:
+  publish_all <tag>
+
+See https://github.com/ChainSafe/lodestar/blob/unstable/RELEASE.md#1-create-release-candidate
+`);
+exitIf(!tag, "<tag> not set");
+
+for (const dirpath of resolveMonorepoDirpaths()) {
+  shell(`npm publish --tag ${tag}`, {cwd: dirpath});
```
Comment (dapplion) [inline scripts/release/publish.mjs:16]: `npm publish` is very simple, just tar files and upload. No surprises.

**Comment by wemeetagain** [comment]
Discussion in discord:
- We'll update this to be in line with https://github.com/ChainSafe/engineering-wiki/pull/28
- If someone on the team takes this task, they should assign themselves first so no one else duplicates their work

**Comment by dapplion** [comment]
Putting as draft until:
- We'll update this to be in line with https://github.com/ChainSafe/engineering-wiki/pull/28

**Comment by dapplion** [comment]
After discussions with the team, we will stick with Lerna for now

**Pull request closed without merging**